Use best practices for wandb.init() and the Run object (#1146)
wandb/docs: content/guides/models/track/config.md

---
{{< cta-button colabLink="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-log/Configs_in_W%26B.ipynb" >}}

> Use the `config` property of a Run to save your training configuration such as: 

Reviewer: ```suggestion
Use the `config` property of a run to save your training configuration: 
```

- hyperparameter
- input settings such as the dataset name or model type
- any other independent variables for your experiments. 

> The `run.config` property makes it easy to analyze your experiments and reproduce your work in the future. You can group by configuration values in the W&B App, compare the settings of different W&B Runs and view how different training configurations affect the output. A Run's `config` property is a dictionary-like object, and it can be built from lots of dictionary-like objects.

Reviewer: ```suggestion
The `run.config` property makes it easy to analyze your experiments and reproduce your work in the future. You can group by configuration values in the W&B App, compare the settings of different W&B runs, and evaluate how each training configuration affects the output. The `config` property is a dictionary-like object that can be composed from multiple dictionary-like objects.
```

> The `run.config` property makes it easy to analyze your experiments and reproduce your work in the future. You can group by configuration values in the W&B App, compare the settings of different W&B Runs and view how different training configurations affect the output. A Run's `config` property is a dictionary-like object, and it can be built from lots of dictionary-like objects.

Reviewer 2: > compare the settings of different W&B Runs
instead
compare the configuration of different W&B Runs

Reviewer 2: avoid confusion with wandb.Settings?

Author: updated!

{{% alert %}}

> Dependent variables (like loss and accuracy) or output metrics should be saved with `run.log` instead.

Reviewer: ```suggestion
To save output metrics or dependent variables like loss and accuracy, use `run.log` instead of `run.config`.
```

Configurations are typically defined in the beginning of a training script. Machine learning workflows may vary, however, so you are not required to define a configuration at the beginning of your training script.

{{% alert color="secondary" %}}

> We recommend that you avoid using dots in your config variable names. Instead, use a dash or underscore instead. Use the dictionary access syntax `["key"]["foo"]` instead of the attribute access syntax `config.key.foo` if your script accesses `run.config` keys below the root.

Reviewer: ```suggestion
Use dashes (`-`) or underscores (`_`) instead of periods (`.`) in your config variable names.

Use the dictionary access syntax `["key"]["value"]` instead of the attribute access syntax `config.key.value` if your script accesses `run.config` keys below the root.
```
I don't think this needs to be an alert.

Author: Updated and removed the alert.

}

# Pass the config dictionary when you initialize W&B
with wandb.init(project="config_example", config=config) as run:
    ...
```

{{% alert %}}

> You can pass a nested dictionary as the `config`. W&B will flatten the names using dots.

Reviewer: ```suggestion
If you pass a nested dictionary as the `config`. W&B flattens the names using dots.
```
I don't think this needs to be an alert.

Author: Updated & removed the alert.

with wandb.init(project="config_example", config=config) as run:
    # Update config after you initialize W&B
    run.config["dropout"] = 0.2
    run.config.epochs = 4
    run.config["batch_size"] = 32
```

You can update multiple values at a time:

```python
run.config.update({"lr": 0.1, "channels": 16})
```

### Set the configuration after your Run has finished
Use the [W&B Public API]({{< relref "/ref/python/public-api/" >}}) to update your config (or anything else about from a complete Run) after your Run. This is particularly useful if you forgot to log a value during a Run. 

Provide your `entity`, `project name`, and the `Run ID` to update your configuration after a Run has finished. Find these values directly from the Run object itself or from the [W&B App UI]({{< relref "/guides/models/track/workspaces.md" >}}):

```python
with wandb.init() as run:
    api = wandb.Api()

    # You can access attributes on the run object directly.
    username = run.entity
    project = run.project
    run_id = run.id

    # NOTE: api.run returns a different type of object than wandb.init().
    api_run = api.run(f"{username}/{project}/{run_id}")
    api_run.config["bar"] = 32

>     api_run.update()

Reviewer 2: this example seems wrong. based on the text above?

Author: can you clarify?

Reviewer 2: oh yeah sure, that we have the context manager to create a run, but the comment above explains how to modify the data post run.

> Provide your `entity`, `project name`, and the `Run ID` to update your configuration after a Run has finished

Author: updated!

```

### Set the configuration after your Run has finished
Use the [W&B Public API]({{< relref "/ref/python/public-api/" >}}) to update a completed run's config. 

> You will need to provide your entity, project name and the run's ID, which you can get from the Run object or find in the [W&B App UI]({{< relref "/guides/models/track/workspaces.md" >}}):

Reviewer: ```suggestion
You must provide the API with your entity, project name and the run's ID. You can find these details in the Run object or in the [W&B App UI]({{< relref "/guides/models/track/workspaces.md" >}}):
```
@ngrayluna Do we say "W&B Public API" the whole thing, with all those caps? Similar quesation for W&B App UI.

Author: Updated!

```python
with wandb.init() as run:
    ...

> # You can access attributes from the run object directly if you are in the same

Reviewer: ```suggestion
# Find the following values from the Run object if it was
#   initiated from the current script or notebook, 
```

Author: Done.

> # script or notebook, or you can get them from the app.

Reviewer: ```suggestion
#   or you can copy them from the App UI.
```

Author: Done.

username = run.entity
project = run.project
run_id = run.id

> # NOTE: api.run returns a different type of object than wandb.init().

Reviewer: ```suggestion
# Note that the api.run() method returns a different type of object than wandb.init().
```
It's not actually an admonition, it's a comment :)

Author: Updated, but FWIW the `NOTE: ...` format is common in code comments. GitHub and VSCode both highlight the "NOTE" so that it's more visible.

    params=hyperparameter_defaults,
)

with wandb.init(config=config_dictionary) as run:
    ...
```

## TensorFlow v1 flags

You can pass TensorFlow flags into the `wandb.config` object directly.

```python
with wandb.init() as run:

>     run.config.epochs = 4

Reviewer: Here are some more literal tabs.

Author: I think these are sequences of 4 spaces, not tab characters. They seem to be rendered correctly:

<img width="918" alt="Screenshot 2025-03-07 at 1 42 26 PM" src="https://github.com/user-attachments/assets/27fcdac6-dcd3-4188-bf2a-800cb682e3b0" />

Am I missing something?

Reviewer: Oh I see now, I was interpreting the diff incorrectly.
